### DISCUSSION TITLE
Need Help! ESP32+LAN8710+Mega2560 OTA update using Serial3

### DISCUSSION THREAD

**Discussion by mikeme2009** [JAndrassy/ArduinoOTA#254] [General]
Hello,

I have a project that contains esp32+mega2560+lan8710. A code runs on the esp32 that does certain things, and another code runs on the Mega, and the two boards communicate with each other on Serial2(Esp32) and Serial3(Mega2560). I need to update Mega using esp32.
On Esp32 I use these 2 functions to download the firmware for Mega2560 from my server, store in SPIFFS and then send over Serial2 to InternalStorage of Mega.

On Mega2560:

#include <ArduinoOTA.h>

#define SERIAL_PORT Serial3

void setup() {
  SERIAL_PORT.begin(115200);  // Ensure this matches the baud rate of your ESP32
  Serial.begin(115200);
}

void loop() {
  if (SERIAL_PORT.available()) {
    String command = SERIAL_PORT.readStringUntil('\n');
    if (command == "START") {
      size_t firmwareSize = SERIAL_PORT.parseInt();  // Read firmware size
      if (firmwareSize > 0) {
        Serial.println("Receiving firmware update");
        receiveFirmware(firmwareSize);
      }
    }
  }
}

void receiveFirmware(size_t length) {
  if (!InternalStorage.open(length)) {
    Serial.println("There is not enough space to store the update. Can't continue with update.");
    return;
  }

  byte b;
  while (length > 0) {
    if (!SERIAL_PORT.readBytes(&b, 1)) {  // Reading a byte with timeout
      break;
    }
    InternalStorage.write(b);
    length--;
  }
  InternalStorage.close();

  if (length > 0) {
    Serial.print("Timeout downloading update file at ");
    Serial.print(length);
    Serial.println(" bytes. Can't continue with update.");
    return;
  }

  Serial.println("Sketch update apply and reset.");
  Serial.flush();
  InternalStorage.apply();  // This doesn't return
}

The problem is when I compile the code I get this error:
FQBN: my_boards:avr:mega
Using board 'mega' from platform in folder: C:\Users\....\Documents\Arduino\hardware\my_boards\avr
...........
:\Users\...\Documents\Arduino\libraries\ArduinoOTA\src/ArduinoOTA.h:141:22: error: 'WiFiServer' was not declared in this scope
 ArduinoOTAMdnsClass <WiFiServer, WiFiClient, WiFiUDP> ArduinoOTA;
                      ^~~~~~~~~~
c:\Users\.......\Documents\Arduino\libraries\ArduinoOTA\src/ArduinoOTA.h:141:22: note: suggested alternative: 'WiFiDrv'
 ArduinoOTAMdnsClass <WiFiServer, WiFiClient, WiFiUDP> ArduinoOTA;
                      ^~~~~~~~~~
                      WiFiDrv
c:\Users\......\Documents\Arduino\libraries\ArduinoOTA\src/ArduinoOTA.h:141:34: error: 'WiFiClient' was not declared in this scope
 ArduinoOTAMdnsClass <WiFiServer, WiFiClient, WiFiUDP> ArduinoOTA;
                                  ^~~~~~~~~~
c:\Users\.......\Documents\Arduino\libraries\ArduinoOTA\src/ArduinoOTA.h:141:34: note: suggested alternative: 'DNSClient'
 ArduinoOTAMdnsClass <WiFiServer, WiFiClient, WiFiUDP> ArduinoOTA;
                                  ^~~~~~~~~~
                                  DNSClient
c:\Users\........\Documents\Arduino\libraries\ArduinoOTA\src/ArduinoOTA.h:141:53: error: template argument 1 is invalid
 ArduinoOTAMdnsClass <WiFiServer, WiFiClient, WiFiUDP> ArduinoOTA;
                                                     ^
c:\Users\.......\Documents\Arduino\libraries\ArduinoOTA\src/ArduinoOTA.h:141:53: error: template argument 2 is invalid

In my code I don't need WiFiServer, WiFiClient, WiFiUDP, but the library requires them.

**Comment by JAndrassy** [comment]
the Advanced examples have
```
#define NO_OTA_NETWORK
#include <ArduinoOTA.h> // only for InternalStorage
```

**Reply by mikeme2009** [reply]
Thanks

**Comment by mikeme2009** [comment]
Hello,
I manage to send the new firmware from ESP32 to Mega using your code with some serial.print for debugging :
void receiveFirmware(size_t length) {
  Serial.println(length);
  if (!InternalStorage.open(length)) {
    Serial.println("There is not enough space to store the update. Can't continue with update.");
    return;
  }

  byte b;
  while (length > 0) {
    if (!Serial3.readBytes(&b, 1)) {  // Reading a byte with timeout
      break;
    }
    InternalStorage.write(b);
    length--;
    Serial.println(length);
  }
  InternalStorage.close();

  if (length > 0) {
    Serial.print("Timeout downloading update file at ");
    Serial.print(length);
    Serial.println(" bytes. Can't continue with update.");
    return;
  }

  Serial.println("Sketch update apply and reset.");
  Serial.flush();
  InternalStorage.apply();  // This doesn't return
}
but in the end the new firmware does not run and the Mega is blocked, the reset button does not work either.
Below is the serial.print. 
........
19:51:00.710 -> 10
19:51:00.710 -> 9
19:51:00.881 -> 8
19:51:01.108 -> 7
19:51:01.108 -> 6
19:51:01.108 -> 5
19:51:01.108 -> 4
19:51:01.108 -> 3
19:51:01.108 -> 2
19:51:01.108 -> 1
19:51:01.108 -> 0
19:51:01.108 -> Sketch update apply and reset.
I check the size of new fw and send this value to mega. I compiled the new sketch with Mega2560(Optiboot) and I tried with both bin files. (with bootloader and without) 
Can you advise me what to check to solve the problem?

**Comment by JAndrassy** [comment]
did you use bin file? (without bootloader of course. only ISP can flash a hex with bootloader)

**Reply by mikeme2009** [reply]
I use bin file. I tried with a small one 5KB. I send the firmware from esp like this:
while (firmwareFile.available()) {
    uint8_t buf[16];
    size_t len = firmwareFile.read(buf, sizeof(buf));
    Serial2.write(buf, len);
    delay(100);  // Small delay to prevent overflow
    Serial.printf("Sent %d bytes to Arduino\n", len);  // Debugging line
  }
The baud rate on Serial2 is 115200

**Comment by JAndrassy** [comment]
if you do a manual reset after OTA update, does the sketch start?

**Reply by mikeme2009** [reply]
nothing happens. I also tried with power off

**Reply by mikeme2009** [reply]
I copied your my_board folder in C:\Users\.......\Documents\Arduino\hardware and copied platform.local.txt from extras/avr folder in Documents\Arduino\hardware\my_board and also to C:\Users\.....\AppData\Local\Arduino15\packages\arduino\hardware\my_boards\avr (I wasn't sure where to copy :( ) then I burned the bootloader on the Mega2560. Did something else need to be done?

**Reply by JAndrassy** [reply]
for InternalStorage that is all.

**Reply by mikeme2009** [reply]
I also did some tests with the bootloader from MegaCore. I burned this bootloader on the Mega 2560 and then compiled the program (to display "Hello world" on the I2C LCD) using the Mega2560(optiboot). After it finished loading the new sketch, the Mega2560 restarted, but instead of displaying "Hello World" on the LCD it appeared on Serial.print! After the physical restart, the board started with the old software.
